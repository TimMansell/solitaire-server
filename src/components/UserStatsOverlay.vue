<template>
  <GameOverlay alt data-test="user-stats-overlay">
    <template #title> Stats </template>
    <template #msg>
      <div>Played: <Counter :number="completed" /> (1 in progress)</div>
      <div>Won: <Counter :number="won.count" /> ({{ won.percent }})</div>
      <div>Lost: <Counter :number="lost.count" /> ({{ lost.percent }})</div>
      <div>Gave up: <Counter :number="abandoned.count" /> ({{ abandoned.percent }})</div>
    </template>
    <template #buttons>
      <Button @click="closeStats" data-test="close-stats-btn">Close</Button>
    </template>
  </GameOverlay>
</template>

<script>
import numeral from 'numeral';
import { mapActions, mapGetters } from 'vuex';
import GameOverlay from '@/components/GameOverlay.vue';
import Button from '@/components/Button.vue';
import Counter from '@/components/Counter.vue';

const calcPercent = (value) => numeral(value).format('0.00%');

export default {
  name: 'GameLost',
  components: {
    GameOverlay,
    Button,
    Counter,
  },
  computed: {
    ...mapGetters(['fullUserStats']),
    completed() {
      const { completed } = this.fullUserStats;

      return completed;
    },
    won() {
      const { completed, won } = this.fullUserStats;
      const percent = calcPercent(won / completed);

      return {
        count: won,
        percent,
      };
    },
    lost() {
      const { completed, lost } = this.fullUserStats;
      const percent = calcPercent(lost / completed);

      return {
        count: lost,
        percent,
      };
    },
    abandoned() {
      const { won, lost, completed } = this.fullUserStats;
      const abandoned = completed - won - lost;
      const percent = calcPercent(abandoned / completed);

      return {
        count: abandoned,
        percent,
      };
    },
  },
  mounted() {
    this.getUserStats();
  },
  methods: {
    ...mapActions(['getUserStats', 'toggleStats', 'setGamePaused']),
    closeStats() {
      const isPaused = {
        isPaused: false,
        isActive: false,
        showMsg: false,
      };

      this.setGamePaused(isPaused);
      this.toggleStats();
    },
  },
};
</script>
