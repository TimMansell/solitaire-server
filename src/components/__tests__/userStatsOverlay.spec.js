import { shallowMount } from '@vue/test-utils';
import UserStatsOverlay from '@/components/UserStatsOverlay.vue';

const mocks = {
  $store: { dispatch: jest.fn() },
};

describe('UserStatsOverlay.vue', () => {
  it('matches snapshot', () => {
    const wrapper = shallowMount(UserStatsOverlay, {
      computed: {
        fullUserStats: () => ({
          count: 4,
          won: 1,
          lost: 1,
          completed: 2,
        }),
      },
      mocks,
    });

    expect(wrapper).toMatchSnapshot();
  });
});
