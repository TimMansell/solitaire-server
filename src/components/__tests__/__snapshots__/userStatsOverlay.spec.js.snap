// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserStatsOverlay.vue matches snapshot 1`] = `
<gameoverlay-stub alt="true" data-test="user-stats-overlay"> Stats <div>Played: <counter-stub number="2"></counter-stub> (1 in progress)</div>
  <div>Won: <counter-stub number="1"></counter-stub> (50.00%)</div>
  <div>Lost: <counter-stub number="1"></counter-stub> (50.00%)</div>
  <div>Gave up: <counter-stub number="0"></counter-stub> (0.00%)</div>
  <button-stub data-test="close-stats-btn">Close</button-stub>
</gameoverlay-stub>
`;
